### PR TITLE
feat: SoftMatcher and PerDINO integration

### LINKED ISSUE #1047: [Task]: Model Contract - Feature Matching Models (Matcher + PerDino + SoftMatcher)
### Description

Rewrite visual feature matching models to use new `TorchModel` base class.
#### 1. Matcher (`library/src/instantlearn/models/matcher/matcher.py`) — REPLACE

```python
class Matcher(TorchModel):
    def __init__(
        self,
        sam: str | Path = ...,
        encoder_model: str = ...,
        num_foreground_points: int = 12,
        num_background_points: int = 0,
        confidence_threshold: float = 0.38,
        precision: str = "fp16",
        compile_models: bool = False,
        device: str = "cuda",
        preprocessor: Preprocessor | None = None,
        postprocessor: PostProcessor | None = None,
    ): ...

    @classmethod
    def card(cls) -> ModelCard:
        return ModelCard(
            name="Matcher",
            family="matcher",
            description="One-shot segmentation via visual feature matching",
            prompt_types=frozenset({PromptType.VISUAL}),
            shot_modes=frozenset({ShotMode.ONE_SHOT, ShotMode.FEW_SHOT}),
            exportable_to=frozenset({Backend.OPENVINO, Backend.ONNX}),
        )

    def fit(self, reference: Sample | list[Sample]) -> None:
        """Extract and cache reference features. Raises nothing — always required."""

    def predict(self, target: Sample | list[Sample]) -> list[Prediction]:
        """Raises ModelNotFittedError if fit() not called."""
```

**Key changes:**

- `predict()` raises `ModelNotFittedError` if called before `fit()`
- `fit()` accepts numpy `Sample` → internally converts to tensors
- `predict()` returns `list[Prediction]` not `list[dict[str, torch.Tensor]]`
- Internal `BidirectionalPromptGenerator`, `MaskedFeatureExtractor`, `ReferenceFeatures` **stay as nn.Module**
- `export()` and `to_openvino()` implemented (current ONNX export logic preserved)
- `prepare_openvino_input()` stays as internal helper

#### 2. PerDino (`library/src/instantlearn/models/per_dino/per_dino.py`) — REPLACE

```python
class PerDino(TorchModel):
    def __init__(self, sam, encoder_model, num_foreground_points, num_background_points,
                 num_grid_cells, point_selection_threshold, confidence_threshold,
                 precision, compile_models, device, preprocessor, postprocessor): ...

    @classmethod
    def card(cls) -> ModelCard: ...  # VISUAL prompt, ONE_SHOT/FEW_SHOT

    def fit(self, reference: Sample | list[Sample]) -> None: ...
    def predict(self, target: Sample | list[Sample]) -> list[Prediction]: ...
```

Same contract. `ModelNotFittedError` if predict before fit.

#### 3. SoftMatcher (`library/src/instantlearn/models/soft_matcher/soft_matcher.py`) — REPLACE

```python
class SoftMatcher(Matcher):  # inherits from new Matcher(TorchModel)
    def __init__(self, ..., softmatching_score_threshold, softmatching_bidirectional, ...): ...
```

Inherits from new `Matcher`. Overrides prompt generator initialization.

### Files to modify

| File | Action |
|------|--------|
| `library/src/instantlearn/models/matcher/matcher.py` | Replace |
| `library/src/instantlearn/models/per_dino/per_dino.py` | Replace |
| `library/src/instantlearn/models/soft_matcher/soft_matcher.py` | Replace |
| `library/src/instantlearn/models/matcher/__init__.py` | Update |
| `library/src/instantlearn/models/per_dino/__init__.py` | Update |
| `library/src/instantlearn/models/soft_matcher/__init__.py` | Update |
| `library/src/instantlearn/models/matcher/prompt_generators.py` | **NO CHANGE** |

### Component

Library (instantlearn Framework)

### Acceptance Criteria

- [ ] All 3 models inherit from `TorchModel`
- [ ] `predict()` returns `list[Prediction]`
- [ ] `predict()` without prior `fit()` raises `ModelNotFittedError`
- [ ] `fit()` accepts numpy `Sample | list[Sample]`
- [ ] Internal components (prompt generators, feature extractors, SAM decoder) untouched
- [ ] `Matcher.card()` returns correct `ModelCard`
- [ ] `SoftMatcher` still inherits from `Matcher`
- [ ] Pass `ruff check`

### Implementation Notes

_No response_

### Parent Epic

#1001

### Dependencies

_No response_

### PR DESCRIPTION
**Pull request by LenaRudaya**
This PR migrates PerDINO and SoftMatcher further into the backend-agnostic model stack, adds baked-export support (ONNX/OpenVINO) with shared export utilities, and standardizes exported artifact naming to model.* across Torch/OpenVINO paths and tests.

- Add/extend baked export pipelines (Torch → ONNX → OpenVINO IR) for PerDINO/SoftMatcher and introduce their OpenVINO “sibling” inference loaders.
- Add ONNX-export-safe prompt-selection branches for PerDINO (grid) and SoftMatcher (soft matching) and tests covering determinism/static shapes.
- Standardize IR/ONNX output names to model.xml/.bin/.onnx, plus metadata handling and various test/notebook updates.

## Type of Change

- [ ] ✨ `feat` - New feature
- [ ] 🐞 `fix` - Bug fix
- [ ] 📚 `docs` - Documentation
- [ ] ♻️ `refactor` - Code refactoring
- [ ] 🧪 `test` - Tests
- [ ] 🔧 `chore` - Maintenance

## Related Issues

Closes #1047

### CODE REVIEW

File: library/src/instantlearn/data/base/sample.py [hunk 1]
```diff
@@ -80,13 +85,27 @@ class Sample:
...
+        """
+        # Normalize mask_paths: str -> list[str]
+        if isinstance(self.mask_paths, str):
+            self.mask_paths = [self.mask_paths]
+
         if self.image is None and self.image_path is not None:
             self.image = read_image(self.image_path)
 
+        if self.masks is None and self.mask_paths is not None:
+            self.masks = np.stack([read_mask(p) for p in self.mask_paths], axis=0)  # (N, H, W)
```
[the remark this answers is not in the document]
Comment (LenaRudaya) [reply library/src/instantlearn/data/base/sample.py:107]: not true, only `Category` dataclass is frozen

File: library/src/instantlearn/components/feature_extractors/masked_feature_extractor.py [hunk 2]
```diff
@@ -95,7 +95,12 @@ def forward(
                     per_image_cat_masks[cat_id] = mask
 
             for cat_id, merged_mask in per_image_cat_masks.items():
-                pooled_mask = self.transform(merged_mask).to(embedding.device)
+                # Binarize the pooled mask to a strict 0/1 tensor. ``ToTensor`` rescales uint8 numpy masks
+                # by 1/255 (torchvision behaviour), so a numpy ``Sample.mask`` yields foreground values of ~0.0039
+                # instead of 1.0. Downstream consumers that threshold at ``> 0.5`` (e.g. the export prompt path)
+                # would then see no foreground and emit degenerate prompts.
+                # Normalizing here keeps the reference masks binary regardless of the source dtype (numpy or torch).
+                pooled_mask = (self.transform(merged_mask).to(embedding.device) > 0).to(embedding.dtype)
```
Comment (rajeshgangireddy) [inline library/src/instantlearn/components/feature_extractors/masked_feature_extractor.py]: Why is this ` > 0` needed now? 
Also, are we sure everything works (especially thresholding)?

**Review by rajeshgangireddy — APPROVED** [review]
Just a small question. Otherwise, it looks good.

File: library/src/instantlearn/components/feature_extractors/masked_feature_extractor.py (hunk 2 above)
Comment (LenaRudaya) [reply library/src/instantlearn/components/feature_extractors/masked_feature_extractor.py]: The `> 0` was re-binarizing the pooled mask. Reference masks now come in as numpy uint8 (from torch-free `Sample.mask`), and torchvision ToTensor rescales those by 1/255 - so foreground 1 became ~0.0039. MatcherOpenVINO thresholds at > 0.5, do it saw no foreground, and predicted garbage (see screenshot - sky and random ground noise instead of the elephants).
image

The > 0 fixed it but was a bit of a band-aid. [I've replaced it](https://github.com/open-edge-platform/geti-instant-learn/pull/1102/changes/27ecb276dd2777439315cca211ed02b8c0b8033c) to be binary by construction instead: convert masks to boolean tensors at ingestion and use nearest-neighbor resize (so edges stay 0/1).

Comment (rajeshgangireddy) [reply library/src/instantlearn/components/feature_extractors/masked_feature_extractor.py]: if the final masks look as expected, then it's a great fix. Thanks.
